ic-task-scheduler/src/scheduler.rs

```diff
@@ -45,22 +55,23 @@ impl<T: 'static + Task, P: 'static + UnboundedMapStructure<u32, ScheduledTask<T>
                         task_execution_started += 1;
                         let task_scheduler = self.clone();
                         Self::spawn(async move {
-                            match task.task.execute(Box::new(task_scheduler.clone())).await {
-                                Ok(()) => {
-                                    // processed_tasks.push(task);
+                            if let Err(err) =
+                                task.task.execute(Box::new(task_scheduler.clone())).await
+                            {
+                                let mut task = task;
+                                task.options.failures += 1;
+                                let (should_retry, retry_delay) = task
+                                    .options
+                                    .retry_strategy
+                                    .should_retry(task.options.failures);
+                                if should_retry {
+                                    task.options.execute_after_timestamp_in_secs =
+                                        now_timestamp_secs + (retry_delay as u64);
+                                    task_scheduler.append_task(task)
                                 }
-                                Err(_) => {
-                                    let mut task = task;
-                                    task.options.failures += 1;
-                                    let (should_retry, retry_delay) = task
-                                        .options
-                                        .retry_strategy
-                                        .should_retry(task.options.failures);
-                                    if should_retry {
-                                        task.options.execute_after_timestamp_in_secs =
-                                            now_timestamp_secs + (retry_delay as u64);
-                                        task_scheduler.append_task(task)
-                                    }
+                                // send error to callback
+                                if let Some(cb) = &*task_scheduler.failed_task_cb {
```

Comment:
The callback should be called only if the task has no more retries.
Add these tests:
- Give a task, when it succeeds, a failure callback should not be called
- Give a task, when it fails and has no retries, a failure callback should be called
- Give a task, when it fails n times but the (n+1) it succeeds, a failure callback should not be called
- Give a task, when it fails n times, a failure callback should be called only once after the last failure